text/0062-controlling-pii-and-credentials-in-sd-ks.md

```diff
@@ -0,0 +1,207 @@
+  * this will be partially fixed by `before_send_transaction` but I still believe PII scrubbing is a key concern that should not be shoe-horned into catch-all callbacks that run at the end
+* writing either of these callbacks requires the user to know the whole event schema which even our own SDK devs don't know fully
+
+# Options Considered
+
+The high-level proposal is to introduce a new `config.event_scrubber` with a default implementation that users can tweak with a simple `denylist` in most cases.  
+The end-user only cares about hiding certain information quickly, so a single list is the ideal interface.  
+Anything more complex than that is asking them to think about things they don't care about.  
+The advanced user can still override the implementation (similar to custom transports) if they wish to do so by using their own class implementation.
+
+
+```python
+# ===========================================================================
+# event_scrubber.py
+
+DEFAULT_DENYLIST = ["password", "HTTP_COOKIE"]
+
+class EventScrubber:
+    def __init__(self, denylist=None):
+        self.denylist = DEFAULT_DENYLIST if denylist is None else denylist
+
+    def scrub_event(self, event):
+        # type: (Event) -> Event
+        # ... scrubbing logic discussed below
+        return event
+
+
+# ===========================================================================
+# client.py
+
+def capture_event(self, event):
+    # ... existing logic
+    if not config.send_default_pii:
+        scrubbed_event = event_scrubber.scrub_event(event)
+
+    new_event = before_send(scrubbed_event)
+    # ... existing logic
+
+
+# ===========================================================================
+# usage
+
+sentry_sdk.init(
+    send_default_pii=False,
+    event_scrubber=EventScrubber(denylist=["custom_sensitive_field"]),  # has default impl
+)
+```
+
+The actual scrubbing logic can be implemented in two ways with trade-offs outlined below.
+
+## A: Walk whole tree
+
+```python
+class EventScrubber:
+    def scrub_list(self, l):
+        for i, v in enumerate(l):
+            if isinstance(v, dict):
+                l[i] = self.scrub_dict(v)
+            elif isinstance(v, list):
+                l[i] = self.scrub_list(v)
+          return l
+
+    def scrub_dict(self, d):
+        for k, v in d.items():
+            if k in self.denylist:
+                d[k] = "[Filtered]"  # Use `AnnotatedValue` here for _meta
+            elif isinstance(v, dict):
+                d[k] = self.scrub_dict(v)
+            elif isinstance(v, list):
+                d[k] = self.scrub_list(v)
+        return d
+
+    def scrub_event(self, event):
+        return self.scrub_dict(event)
+```
+
+Here we would walk the whole event dict and whenever we encounter a key from the denylist, we would replace the value with `[Filtered]`.
```

Comment:
Could false positives bite use here? I might want to filter a property with the same name in HTTP headers but not in some other `data` field or similar.